File: codeForces/contests/954div3/E.py

from collections import defaultdict
for _ in range(int(input())):
    n, k = list(map(int, input().split()))
    a = list(map(int, input().split()))
    d = defaultdict(list)
    for e in a:
        d[e%k].append(e)
    ans = 0
    senar = False
    for v in d.values():
        v.sort()
        if len(v)%2:
            if n%2 and not senar:
                senar = True
                if len(v) == 1:
                    continue
                dp = [0, v[1]-v[0]]
                for i in range(2, len(v)):
                    dp.append(v[i]-v[i-1]+dp[i-2])
                    if not i%2 and dp[i-1] < dp[-1]:
                        dp[-1] = dp[i-1]
                ans += dp[-1] // k
                continue
            else:
                ans = -1
                break
        for i in range(len(v)//2):
            ans += (v[2*i+1] - v[2*i]) // k
    print(ans)

File: codeForces/contests/954div3/E.cpp
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>

using namespace std;

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, k;
        cin >> n >> k;
        vector<int> a(n);
        map<int, vector<int>> d;
        for (int i = 0; i < n; ++i) {
            cin >> a[i];
            d[a[i] % k].push_back(a[i]);
        }
        int ans = 0;
        bool senar = false;
        for (auto &kv : d) {
            vector<int> &v = kv.second;
            sort(v.begin(), v.end());
            if (v.size() % 2 != 0) {
                if (n % 2 != 0 && !senar) {
                    senar = true;
                    if (v.size() == 1) {
                        continue;
                    }
                    vector<int> dp = {0, v[1] - v[0]};
                    for (int i = 2; i < v.size(); ++i) {
                        dp.push_back(v[i] - v[i - 1] + dp[i - 2]);
                        if (i % 2 == 0 && dp[i - 1] < dp.back()) {
                            dp.back() = dp[i - 1];
                        }
                    }
                    ans += dp.back() / k;
                    continue;
                } else {
                    ans = -1;
                    break;
                }
            }
            for (int i = 0; i < v.size() / 2; ++i) {
                ans += (v[2 * i + 1] - v[2 * i]) / k;
            }
        }
        cout << ans << endl;
    }
    return 0;
}